// react-image-resizer/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Image = function (_React$Component) {
  _inherits(Image, _React$Component);

  function Image(props) {
    _classCallCheck(this, Image);

    var _this = _possibleConstructorReturn(this, (Image.__proto__ || Object.getPrototypeOf(Image)).call(this, props));

    _this.state = {
      width: 0,
      height: 0,
      isNoImage: false
    };
    _this.resizeImage = _this.resizeImage.bind(_this);
    _this.showNoImage = _this.showNoImage.bind(_this);
    return _this;
  }

  _createClass(Image, [{
    key: 'resizeImage',
    value: function resizeImage() {
      if (_reactDom2.default.findDOMNode(this.refs.image) == null) {
        return;
      }
      var originalWidth = _reactDom2.default.findDOMNode(this.refs.image).naturalWidth;
      var originalHeight = _reactDom2.default.findDOMNode(this.refs.image).naturalHeight;
      var widthRatio = this.props.width / originalWidth;
      var heightRatio = this.props.height / originalHeight;
      if (widthRatio < heightRatio) {
        this.setState({
          width: originalWidth * widthRatio,
          height: originalHeight * widthRatio
        });
      } else {
        this.setState({
          width: originalWidth * heightRatio,
          height: originalHeight * heightRatio
        });
      }
    }
  }, {
    key: 'showNoImage',
    value: function showNoImage() {
      if (this.props.noImageSrc == undefined) {
        return;
      }
      this.setState({
        isNoImage: true
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var style = {
        wrapper: {
          position: 'relative',
          width: this.props.width,
          height: this.props.height,
          backgroundColor: this.props.backgroundColor
        },
        image: {
          position: 'absolute',
          display: 'block',
          left: (this.props.width - this.state.width) / 2,
          top: (this.props.height - this.state.height) / 2,
          width: this.state.width,
          height: this.state.height
        }
      };
      if (this.state.isNoImage) {
        return _react2.default.createElement(
          'div',
          { style: (0, _utils.m)(this.props.style, style.wrapper) },
          _react2.default.createElement('img', {
            ref: 'image', src: this.props.noImageSrc, alt: this.props.noImageAlt || 'noimage', style: style.image,
            onLoad: this.resizeImage
          })
        );
      } else {
        return _react2.default.createElement(
          'div',
          { style: (0, _utils.m)(this.props.style, style.wrapper) },
          _react2.default.createElement('img', {
            ref: 'image', src: this.props.src, alt: this.props.alt, style: style.image,
            onLoad: this.resizeImage,
            onError: this.showNoImage
          })
        );
      }
    }
  }]);

  return Image;
}(_react2.default.Component);

exports.default = Image;


Image.PropTypes = {
  src: _propTypes2.default.string.isRequired,
  width: _propTypes2.default.number.isRequired,
  height: _propTypes2.default.number.isRequired,
  alt: _propTypes2.default.string,
  backgroundColor: _propTypes2.default.string,
  style: _propTypes2.default.object,
  noImageSrc: _propTypes2.default.string,
  noImageAlt: _propTypes2.default.string
};
